server/src/main/java/org/elasticsearch/snapshots/InvalidSnapshotNameException.java

```diff
@@ -18,17 +22,63 @@
  */
 public class InvalidSnapshotNameException extends SnapshotException {
 
-    public InvalidSnapshotNameException(final String repositoryName, final String snapshotName, String desc) {
-        super(repositoryName, snapshotName, "Invalid snapshot name [" + snapshotName + "], " + desc);
+    @Nullable
+    private final Reason reason;
+
+    public InvalidSnapshotNameException(final String repositoryName, final String snapshotName, Reason reason) {
+        super(repositoryName, snapshotName, "Invalid snapshot name [" + snapshotName + "], " + reason.description);
+        this.reason = reason;
     }
 
     public InvalidSnapshotNameException(StreamInput in) throws IOException {
         super(in);
+        if (in.getVersion().onOrAfter(Version.V_8_2_0)) {
+            reason = in.readOptionalEnum(Reason.class);
+        } else {
+            reason = null;
```

Comment:
What about having something like `reason = Reason.UNKNOWN` rather than `reason = null`? `UNKNOWN` could be documented as being just for backward compatibility purposes (you could even validate that it's not permitted in the non-StreamInput constructor). Then all the `@Nullable`s could drop out. (Note: I'm not even entirely sure I agree with myself on this one -- this is very much "what do you think?".)

Reply:
Interesting, I was reluctant to introduce `Reason.UNKNOWN` because I thought the combination of an optional reason and null it represented better the fact that the reason is missing because of backwards compatibility.

On the other hand, I like the completeness of having `Reason.UNKNOWN` instead of null, especially if you think using a enum with a switch expression.

I am thinking of introducing this and specifying that it is because of bwc. This would limit future abuse of the value and it can be cleaned up when the backwards compatible clause is not needed anymore.